rfd/0163-vnet.md

```diff
@@ -0,0 +1,326 @@
+       its local state.
+    1. VNet will query the cluster(s) apps to see if any have a `spec.vnet_addr`
+       matching the full name. This will use a predicate expression to do the
+       filtering server-side and avoid listing all apps.
+    1. If a match is found a free IP will be assigned to that app and it will be
+       returned in an authoritative answer.
+    1. If no match is found and the zone has no `upstream_nameservers` configured,
+       no answer will be sent.
+    1. The full DNS query will be forwarded to one of the `upstream_nameservers`,
+       selected randomly from the list, and any upstream response will be
+       returned to the client verbatim.
+       We may implement caching of upstream responses, but hopefully the
+       downstream resolver will do its own caching.
+1. If the name did not match any `<proxy-address>.internal` or custom DNS zone,
+   no response will be sent.
+
+TODO: sequence diagram
+
+### IP range
+
+By default IPs will be assigned from the carrier-grade NAT range 100.64.0.0/10
+Since VNet IPs will only be used locally on end user machines this range is
+quite unlikely to conflict with any valid public or private IP addresses, unless
+another VPN client is being used that happens to use the same range.
+For example, tailscale also uses this range.
+This range can be overriden with a cluster-wide configuration.
+
+```yaml
+version: v1
+kind: vnet
+metadata:
+  name: vnet
+spec:
+  cidr_range: 100.64.0.0/10
+```
+
+### Host configuration
+
+Vnet will need to create a TUN interface, assign an IP to it, and modify the
+routing table.
+
+To create the TUN interface we will import `golang.zx2c4.com/wireguard/tun`
+which provides cross-platform methods for creating TUN interfaces and reading
+from + writing them.
+It supports MacOS, Windows, and Linux.
+
+On MacOS, to assign an IP address to the interface, we will use `os/exec` to run
+the `ifconfig` command:
+
+```shell
+ifconfig utun4 100.64.0.1 100.64.0.1 up
+```
+
+On MacOS, to assign an IP route to use the TUN interface, we will use `os/exec`
+to run the `route` command:
+
+```shell
+route add -net 100.64.0.0/10 -interface utun4
+```
+
+All system configuration must run as a privileged user/root.
+In the first version we will use an Applescript wrapper to call a child process
+as root.
+The child admin process will still be the `tsh` binary, just called with
+arguments to do the host setup.
+See the UX section for more details on the admin process.
+
+The first version of VNet will only support MacOS so this is written with that
+in mind, but we can implement this in a similar way for Windows and Linux.
+
+### Security
+
+### UX
+
+VNet is available in both tsh and Teleport Connect. This section focuses mostly on VNet integration
+in Teleport Connect, with a short subsection on VNet in tsh at the end.
+
+Attempting to connect to a TCP app opens the VNet tab instead. When the VNet app is opened this way,
```

Comment:
I'm curious what's going to be in this VNet tab

I've been thinking that there are going to be certain classes of error that we can detect in the VNet server but would be pretty hard to diagnose from the DNS or TCP client side, I'm wondering if we could log these and display recent errors to the user somewhere in Connect. I'm thinking of things like `DNS query for "api.example.private" in custom DNS zone failed: no matching Teleport app and upstream nameserver did not respond` or `failed to dial app.teleport.example.com: some error message`. Even if we just log these to their own file and show the path to it somewhere as a debug option

Reply:
I'm yet to expand this session, but for now I was thinking of start/stop and a basic list of recent connections, sorted by time of last connection attempt.

We could associate the errors you suggested with a specific app. Then with one click you could inspect them and we could also highlight that app in the list if the recent connection resulted in an error.

tsh daemon already captures logs to a file. Initially I was thinking of a separate log file for VNet, for extra logs not associated with a specific app. But I think they can just go to tshd.log.

Connect runs with debug logs enabled btw. We already direct the user to some log files:

* Open the app and then in the menu bar select Help -> Open Logs Directory.
* Set up Connect My Computer, then open the three dots menu in the Connect My Computer tab and select "Open agent logs directory".

Reply:
I think while it would be helfpul to troubleshoot, I'd prefer the UX to make it completely VPN-like (transparent) for user, so they won't have to think about VNet or tunnels 99% of the time.

Reply:
> (…) I'd prefer the UX to make it completely VPN-like (transparent) for user, so they won't have to think about VNet or tunnels lifecycle 99% of the time. (…) We would also de-emphasize the tunneling portion of current Teleport Connect U.X. focusing on virtual IPs and DNS instead.

Yeah, that's what I'm aiming for at the moment. And I agree, once we add support for more protocols, local proxies will become irrelevant.

In the first version though, the VNet likely won't auto start until you start it for the first time. This is because v1 of VNet will ask for root on each start and most users use only ssh and db access – I don't want them to see the root prompt.